e2e/nomostest/config-sync.go

```diff
@@ -1255,96 +1257,147 @@ func resetRepository(nt *NT, repoType RepoType, nn types.NamespacedName, upstrea
 }
 
 func resetRootRepo(nt *NT, rsName, upstream string, sourceFormat filesystem.SourceFormat) {
-	rs := fake.RootSyncObjectV1Beta1(rsName)
-	if err := nt.Get(rs.Name, rs.Namespace, rs); err != nil && !apierrors.IsNotFound(err) {
-		nt.T.Fatal(err)
-	} else {
+	rs := &v1beta1.RootSync{}
+	err := nt.Get(rsName, configmanagement.ControllerNamespace, rs)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			nt.T.Fatal(err)
+		}
+		// If not found, reset the Repository and create the RootSync.
+		// This can happen if it was managed by another RootSync which has
+		// already been reset.
 		nt.RootRepos[rsName] = resetRepository(nt, RootRepo, RootSyncNN(rsName), upstream, sourceFormat)
-		if err == nil {
-			nt.MustMergePatch(rs, fmt.Sprintf(`{"spec": {"sourceFormat": "%s", "git": {"dir": "%s"}}}`, sourceFormat, AcmeDir))
-		} else {
-			rs.Spec.SourceFormat = string(sourceFormat)
-			rs.Spec.Git = &v1beta1.Git{
-				Repo:      nt.GitProvider.SyncURL(nt.RootRepos[rsName].RemoteRepoName),
-				Branch:    MainBranch,
-				Dir:       AcmeDir,
-				Auth:      "ssh",
-				SecretRef: v1beta1.SecretReference{Name: controllers.GitCredentialVolume},
-			}
-			if err = nt.Create(rs); err != nil {
-				nt.T.Fatal(err)
-			}
+		defaultRS := RootSyncObjectV1Beta1FromRootRepo(nt, rsName)
+		if err = nt.Create(defaultRS); err != nil {
+			nt.T.Fatal(err)
 		}
+		return
+	}
+
+	defaultRS := RootSyncObjectV1Beta1FromRootRepo(nt, rsName)
+	if cmp.Equal(rs.Spec, defaultRS.Spec) {
+		// If found and unchanged, just reset the Repository.
+		// This can happen if the RootSync was not modified by the test.
+		nt.RootRepos[rsName] = resetRepository(nt, RootRepo, RootSyncNN(rsName), upstream, sourceFormat)
+		return
 	}
+
+	// If found and needs changes, we need to pause reconciliation to avoid a
+	// race condition between the reconciler-manager and the reconciler.
```

Comment:
What is the race condition?

Reply:
In multi-repo, when you change spec.git.repo or spec.sourceFormat, instead of the reconciler just detecting it, the reconciler-manager detects it (watch on RSync resource) and modifies the reconciler config in its Deployment, causing the reconciler pod to be replaced.

And when you reset the Git repo, adding more commits or changing the branch, the reconciler detects that (git-sync polling git server) and starts a Sync.

So if you change SourceFormat, that requires changing both Git (add or remove the Repo object) and the RSync (spec.sourceFormat). If the reconciler picks up the Git change before the reconciler Pod is stopped by the Deployment change, the reconciler will error, saying either that the Repo is required or not allowed, depending on the direction of the SourceFormat change.

The error should just recover, when the reconciler Pod comes up with a the matching configuration. But for some reason, it's still causing test failure. So I put in these defensive measures to avoid it. Since I had to rewrite a bunch of other reset code to actually get it to work, tho, this may or may not be necessary any more... So I'm doing more testing now.

Reply:
I think the race condition I was trying to solve is actually more of a problem in the mono-repo mode. Unlike multi-repo, there's no reconciler-manager in mono-repo mode, and the config for the SourceFormat is stored in a `source-format` ConfigMap that's loaded into the `importer` container of the `git-importer` Deployment. If the importer picks up the Git change before the ConfigMap has been updated, it can also error. But when there's an error with a specific commit in mono-repo, it gets stuck and stops accepting new commits, causing all subsequent tests to fail.

Reply:
Removed the deployment scale down for now. I don't think multi-repo actually needs it. But mono-repo might.